keps/sig-api-machinery/5366-graceful-leader-transition/README.md

```diff
@@ -0,0 +1,940 @@
+- Document the new development best practices for graceful shutdown requirements
+  for modified components that are leader elected.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The core of this change involves modifying the `OnStoppedLeading` callback to
+not forcefully exit.
+
+We will wrap leader election with a `wait.Until()` to retry the leader election
+loop similar to how the Coordinated Leader Election controller handles
+gracefully transition of leaders
+([code](https://github.com/kubernetes/kubernetes/blob/release-1.33/pkg/controlplane/controller/leaderelection/run_with_leaderelection.go#L54))
+
+The `controller-manager` sets up controller level health checks in
+non-reversible ways and will need to be modified so that handlers can be
+deregistered from the mux when leadership is lost. All resources created after a
+KCM becomes leader must be released when it loses leadership. Some additional
+refactoring may be needed to clean up processes gracefully when a leader lock is
+released.
+
+
+Similarly for scheduler, assumptions that the process will be terminated losing
+the leader lock are made. Many scheduler resources are created before the leader
+election process. These will be modified to either defer resource creation or
+add a resetting mechanism when the leader is lost.
+
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+To properly test that components can respect context cancellation shutdowns and
+not leak memory, we will run both manual tests with profiling, and automated
+testing that transitions the leader lock multiple times.
+
+We will also test that things like caches, reflectors, and go funcs are properly
+cleared/stopped when the leader lock is lost.
+
+Scenarios:
+
+- Component acquires leader lock should start its control loop
+- Component releasing leader lock should shut down its control loop
+- Component acquiring and releasing the leader lock multiple times should not leak memory
+- Component acquiring and releasing the leader lock multiple times should have ONLY ONE control loop running
```

Comment:
on release, we have zero control loops, right?

Reply:
Yes. This case is for on acquire where only one control loop for the controllers should be running (on the active leader instance). This is an invariant that we assume is true, but not well tested.

Reply:
Please explicitly add the test.